docs/Style.thy

```diff
@@ -0,0 +1,182 @@
+                       'a my_very_long_list_type \<Rightarrow> 'a my_very_long_option_type" where
+  "hd_opt2 dummy_list dummy_nat \<equiv> case_list None (\<lambda>h t. Some h)"
+
+(* In the case that the definition body causes an overflow of the 100 char line limit, use one
+   of the following two patterns.
+
+   Note that in each case, indentation of the right-hand expression remains no less than the
+   indentation level of `case_list`.
+ *)
+
+definition hd_opt3 :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt3 \<equiv> case_list
+             None
+             (\<lambda>h t. Some h)"
+
+definition hd_opt4 :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt4 \<equiv>
+   case_list None (\<lambda>h t. Some h)"
+
+
+section \<open>Lemma Statements\<close>
+
+(* Lemmas should be written over 2 lines in the following style.
+
+   Note that indentation is 2 spaces and the lemma statement does not run past 100 character line
+   limit. Note also that multiple premises are listed within `\<lbrakk> \<rbrakk>`.
+
+  For example,
+*)
+
+lemma test_lemma0:
+  "\<lbrakk>hd_opt l \<noteq> None; hd_opt (hd l) = Some x\<rbrakk> \<Longrightarrow>
+   hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                 split: list.splits)
+  done
+
+section \<open>Proofs\<close>
+
+subsection \<open>Line length: 100\<close>
+
+(* The line length should be restricted to 100 characters wherever possible. See below for how to
+   indent statements that need to be broken over multiple lines. *)
+
+subsection \<open>Indentation (general)\<close>
+
+(* From Gerwin's guide:
+
+   If an apply command is applied to a state with n+1 subgoals, it must be indented by n spaces
+   relative to the first apply in the sequence (see the Isabelle distribution for many examples).
+   This indicates textually which line solved a subgoal and how many new subgoals where generated.
+   Invaluable information when something goes wrong because something changed somewhere. Also, it
+   gives a nice visual indication of the nesting depth of your proof.
+
+   The usual indentation for 0 subgoals is 2 spaces for a top-level proof.
+
+  For example:
+*)
+
+lemma test_lemma1:
+  "2 \<le> length l \<Longrightarrow> \<exists>h1 h2 t. l = h1 # h2 # t"
+  apply (case_tac l)
+   apply simp
+  apply (rename_tac h1 t1)
+  apply (case_tac t1)
+   apply simp
+  apply simp
+  done
+
+(* TO DO: Improve this example. *)
+
+subsection \<open>Indentation (multi-line proofs)\<close>
+
+(* If a line of proof will exceed 100 characters then it should be wrapped to the next line and
+   remain indented at a depth no less than it was already.
+
+   TO DO: Develop more rigorous advice on this issue.
+
+   Some examples.
+*)
+
+lemma test_lemma2:
+  "case_option None hd_opt (hd_opt l) = Some x \<Longrightarrow>
+   hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                 split: list.splits)
```

Comment:
It's in the TODO, but just wanted to comment here that we should allow both alignment styles for method modifiers: align on `:` and align left.